### ISSUE TITLE
PGXL parallel dump

### ISSUE THREAD

**Issue by zifeishan** [HazyResearch/deepdive#372]
In PostgresXL, currently there are no parallel dump support, and we are using `COPY` command to dump extractor input.

One way for parallel dump is that: if all the join tables have same distribution key, we can control individual data nodes to do local joins and dump in parallel. The datanodes can be connected in read-only mode through command line, and if there are N data nodes, the dump is essentially N times faster than single-threaded COPY.

**Comment by chrismre** [comment]
hmm... this should be handled by DD... Seems easy enough? @zhangce @feiranwang

**Comment by alldefector** [comment]
Related to this, when loading data, the coordinator is also the bottleneck. We haven't found a good way to do parallel loading yet. PG 9.5 offers the ability to switch b/w LOGGED and UNLOGGED for a table, which could be used to speed up data loading for PG. Unfortunately, PGXL is based on PG 9.2 and so cannot benefit from that.

**Comment by feiranwang** [comment]
Synced up with Zifei. The described approach applies to dumping a single table or local join of tables. In DD, an extractor input query could be an arbitrary SQL query, and the approach could not apply to such a general case. We could optimize for specific types of queries using the ad hoc approach, by getting hints from the users, which I think might be too specific.

**Comment by chrismre** [comment]
Why not make sure the simple cases work? We don't have to handle the
general case to be useful, right?

**Comment by feiranwang** [comment]
Sure, we can add support for that.

**Comment by chrismre** [comment]
Maybe just an annotation?

**Comment by feiranwang** [comment]
Yes, basically our idea is to add parameters from the user to deepdive, and perform local join and dumping.

**Comment by raphaelhoffmann** [comment]
I did some experiments which suggest PGXL unload performance *is* really
good! I'm consistently getting 200MB/sec, which is the max I/O performance
for the machine.

However, I also did experience the slow unload performance Zifei has
reported on a different machine (with the same specs). This must be a PGXL
bug or configuration issue.

What is really slow, however, is loading (25MB/sec) which is CPU-bound and
not parallel.

alldefector: PGXL does support UNLOGGED and we've been using it.

**Comment by badder** [comment]
@raphaelhoffmann : what throughput do you get for loading with UNLOGGED? Did you find a work-around to turn off UNLOGGED after loading?

**Comment by alldefector** [comment]
@raphaelhoffmann : what throughput do you get for loading with UNLOGGED? Did you find a work-around to turn off UNLOGGED after loading?

**Comment by raphaelhoffmann** [comment]
Loading into UNLOGGED table: 100MB/sec
Loading into LOGGED table: 21MB/sec

Postgres 9.5 includes a new command
ALTER TABLE ... SET LOGGED;

But there does not seem to be an option to convert unlogged to logged in
9.2/PGXL.

**Comment by raphaelhoffmann** [comment]
I created a little script that does parallel unload from PGXL, runs a
command, then does parallel loading. It reads the port numbers of the
running PGXL nodes from the coordinator.

https://gist.github.com/raphaelhoffmann/916a97435801d438c919

This should be fast, but I am still planning to do a more careful
performance analysis.

One more thing: On Greenplum, I noticed one can also get a load time
improvement by running multiple \COPY in parallel – however, on raiders4
there's a sweet spot which is somewhere between 2 and 9 parallel processes.
After that, you see a decrease in performance, likely due to cache misses.




On Tue, Sep 1, 2015 at 6:42 PM, Raphael Hoffmann <raphael.hoffmann@gmail.com
